src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -0,0 +1,371 @@
+      // release lock
+      bufferLock.unlock();
+    }
+    // intermediate buffer can be empty here if time interval reached but kafka produced no records.
+    if (intermediateBuffer.isEmpty()) {
+      LOGGER.info("No Rows Buffered, returning;");
+      this.previousFlushTimeStampMs = System.currentTimeMillis();
+      return;
+    }
+    InsertValidationResponse response =
+        channel.insertRows(
+            (Iterable<Map<String, Object>>) intermediateBuffer.getData(),
+            Long.toString(intermediateBuffer.getLastOffset()));
+
+    // need to update the flush time(last I called insertrows)
+    this.previousFlushTimeStampMs = System.currentTimeMillis();
+
+    // should we retry if response.hasErrors() == true?
+    // how can we retry if buffer was cleared?
+    // lets have an exponential back off and jitter retry policy
+
+    if (response.hasErrors()) {
+      LOGGER.error(
+          "Inserted rows API response has errors for rows with offset start:{} and end:{}",
+          intermediateBuffer.getFirstOffset(),
+          intermediateBuffer.getLastOffset());
+      response
+          .getInsertErrors()
+          .forEach(
+              insertError -> {
+                LOGGER.error(
+                    "Insert row Error message:{}, with ex:{}",
+                    insertError.getMessage(),
+                    insertError.getException().getMessage());
+              });
+    }
+  }
+
+  // should we rely on precommits frequency to give us the last committed offset token?
+  // or should we have a BG thread polling on every channel to update the Atomic long and simply
+  // return the atomic long's current value
+  // For now we will rely on preCommit API to fetch the inserted offsets into Snowflake
+  // TODO: SNOW-529755 POLL committed offsets in backgraound thread
+  /**
+   * Get committed offset from Snowflake. It does an HTTP call internally to find out what was the
+   * last offset inserted.
+   *
+   * <p>If committedOffset fetched from Snowflake is null, we would return -1(default value of
+   * committedOffset) back to original call. (-1) would return an empty Map of partition and offset
+   * back to kafka. (We)
+   *
+   * <p>Check {@link com.snowflake.kafka.connector.SnowflakeSinkTask#preCommit(Map)}
+   *
+   * <p>Else, we will convert this offset and return the offset which is safe to commit inside
+   * Kafka.
+   *
+   * @return offsetToken present in Snowflake, else -1
+   */
+  public long getCommittedOffset() {
+    LOGGER.debug(
+        "Fetching last committed offset for partition channel:{}",
+        this.channel.getFullyQualifiedName());
+    String lastOffsetCommitted = channel.getLatestCommittedOffsetToken();
+    LOGGER.info(
+        "Last committed offset for partition channel:{} is :{}",
+        this.channel.getFullyQualifiedName(),
+        lastOffsetCommitted);
+    if (Strings.isNullOrEmpty(lastOffsetCommitted)
+        || lastOffsetCommitted.equalsIgnoreCase("null")) {
```

Comment:
I don't think the lastOffsetCommitted could be "null" string, it should be either valid and NULL

Reply:
Okay makes sense, I confirmed from gs side, we only set a valid one and if it is not set, we return null.

Reply:
I will remove the second condition. Thanks